File: web/src/audio/onset.js
/**
 * Onset detection by spectral flux — the JS side of spike/core/onset.py.
 * See spike/ALGORITHM.md §3.
 *
 * Not optional. Without it a chord still ringing from the previous step
 * satisfies the next step for free and the song fast-forwards through itself
 * while the player sits there. The sustain pedal makes that worse by design.
 */

import { MIN_HISTORY_FRAMES, onsetRefractoryFrames } from './params.js';

/**
 * Sum of positive bin-to-bin magnitude increases.
 * Half-wave rectified: a decaying note produces large negative deltas, and
 * counting those would make one chord's release look like the next's attack.
 */
export function spectralFlux(spec, prev) {
  if (!prev || prev.length !== spec.length) return 0;
  let sum = 0;
  for (let i = 0; i < spec.length; i++) {
    const d = spec[i] - prev[i];
    if (d > 0) sum += d;
  }
  return sum;
}

export class OnsetDetector {
  constructor(params) {
    this.params = params;
    this.refractoryFrames = onsetRefractoryFrames(params);
    this._prev = null;
    this._history = [];
    this._sinceOnset = this.refractoryFrames;
    this.lastFlux = 0;
    this.lastThreshold = 0;
  }

  reset() {
    this._prev = null;
    this._history.length = 0;
    this._sinceOnset = this.refractoryFrames;
    this.lastFlux = 0;
    this.lastThreshold = 0;
  }

  /** Feed one magnitude spectrum; returns whether this frame is an onset. */
  step(spec) {
    this._sinceOnset++;

    const flux = spectralFlux(spec, this._prev);
    if (!this._prev || this._prev.length !== spec.length) {
      this._prev = new Float64Array(spec.length);
    }
    this._prev.set(spec);

    let total = 0;
    for (let i = 0; i < spec.length; i++) total += spec[i];
    const floor = this.params.onsetFluxFloor * total;

    // Median over history EXCLUDING the current frame, so a strong onset does
    // not raise the very bar it has to clear.
    const warmingUp = this._history.length < MIN_HISTORY_FRAMES;
    let threshold;
    if (warmingUp) {
      threshold = floor;
    } else {
      const sorted = Float64Array.from(this._history).sort();
      const mid = sorted.length >> 1;
      const median =
        sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
      threshold = Math.max(this.params.onsetFluxRatio * median, floor);
    }

    this._history.push(flux);
    if (this._history.length > this.params.onsetWindowFrames) this._history.shift();
    this.lastFlux = flux;
    this.lastThreshold = threshold;

    // No onsets until the passage has been observed. Otherwise the first
    // frames — where the only comparison is the absolute floor — report room
    // noise as a strike, arming the verifier before a key is touched.
    if (warmingUp) return false;
    if (flux <= threshold || flux <= 0) return false;

    // Refractory gate. History is still updated above, so a suppressed burst
    // continues to inform the median; only the reported onset is swallowed.
    if (this._sinceOnset < this.refractoryFrames) return false;

    this._sinceOnset = 0;
    return true;
  }
}

File: spike/core/onset.py
"""Onset detection by spectral flux.

Open decision F, and not optional. From PLAN's risk list:

    "Onset gating is required so a still-ringing chord doesn't auto-satisfy the
    next step."

Concretely: the player holds a C major triad, the app advances, and the next
step is C major again (or shares its pitch classes). Without an onset
requirement the still-decaying sound satisfies the new step instantly, the app
advances again, and the song fast-forwards through itself while the player sits
there. Sustain pedal makes this worse by design.

Flux is measured as the sum of positive bin-to-bin magnitude increases between
consecutive frames — energy *appearing*, which is what a hammer strike does,
while decay contributes nothing because those deltas are negative.
"""

from __future__ import annotations

from collections import deque

import numpy as np

from ..params import Params

MIN_HISTORY_FRAMES = 3
"""Frames of flux history required before any onset may be reported."""


def spectral_flux(spec: np.ndarray, prev_spec: np.ndarray | None) -> float:
    """Sum of positive magnitude increases from `prev_spec` to `spec`.

    Half-wave rectified on purpose: a decaying note produces large *negative*
    deltas, and counting those would make the release of one chord look like
    the attack of the next.
    """
    if prev_spec is None or len(prev_spec) != len(spec):
        return 0.0
    return float(np.sum(np.maximum(spec - prev_spec, 0.0)))


class OnsetDetector:
    """Adaptive-threshold onset detector over a rolling median of recent flux.

    A fixed flux threshold cannot work: flux scales with absolute loudness, so
    any constant is simultaneously too high for soft playing and too low for
    loud playing. Comparing against the recent median instead asks the
    scale-free question "is this frame unusually energetic *for this passage*",
    which survives both dynamics and mic gain.
    """

    def __init__(self, params: Params):
        self.params = params
        self._prev_spec: np.ndarray | None = None
        self._history: deque[float] = deque(maxlen=params.onset_window_frames)
        self._since_onset = params.onset_refractory_frames
        self.last_flux = 0.0
        self.last_threshold = 0.0

    def reset(self) -> None:
        self._prev_spec = None
        self._history.clear()
        self._since_onset = self.params.onset_refractory_frames
        self.last_flux = 0.0
        self.last_threshold = 0.0

    def step(self, spec: np.ndarray) -> bool:
        """Feed one frame's magnitude spectrum; return whether it is an onset."""
        self._since_onset += 1
        flux = spectral_flux(spec, self._prev_spec)
        self._prev_spec = spec.copy()

        total = float(np.sum(spec))
        floor = self.params.onset_flux_floor * total

        # Median over history *excluding* the current frame, so a strong onset
        # does not raise the very bar it has to clear.
        warming_up = len(self._history) < MIN_HISTORY_FRAMES
        if warming_up:
            threshold = floor
        else:
            median = float(np.median(self._history))
            threshold = max(self.params.onset_flux_ratio * median, floor)

        self._history.append(flux)
        self.last_flux = flux
        self.last_threshold = threshold

        # No onsets until the passage has been observed for a few frames.
        # Otherwise the very first frames — where the only thing to compare
        # against is the absolute floor — report room noise as a strike, which
        # arms the verifier before a single key is touched and silently
        # cancels the onset gate that decision F depends on.
        if warming_up:
            return False

        if flux <= threshold or flux <= 0.0:
            return False

        # Refractory gate. The flux history is still updated above, so a
        # suppressed burst continues to inform the adaptive median — only the
        # reported onset is swallowed.
        if self._since_onset < self.params.onset_refractory_frames:
            return False

        self._since_onset = 0
        return True
